### ISSUE TITLE
[Collection] Addressing major issues

### ISSUE THREAD

**Issue by TimoBechtel** [TimoBechtel/socketdb#54]
This issue is a non-exhaustive list of issues. Those are major issues that may be related to one another, but in any case require quite a lot of refactoring and/or breaking changes.

---

SocketDB has a few major issues at the moment that need to be addressed:

## Server Logic

The current implementation of the server API is very low level, for example to update state, one has to create a complete update object:

```ts
server.update(nodeify({
	posts: {
		1: {
			likes: 10
		}
	}
}));
```

This is quite cumbersome, especially when trying to update the metadata, because nodeify does not support metadata.

The same applies to plugins, as the `server:update` hook only passes a complete update object, similar to the above.
So to check what has been updated, every plugin has to traverse the whole update object just to get the actual values:

```ts
const myPlugin: ServerPlugin = {
	name: 'myPlugin',
	hooks: {
		'server:update': ({data}) => {
			traverseNode(data, (path, node) => {
				if (!hasChildNodes(node)) {
					// this is the updated leaf value
				}
			});
		}
	}
}
```

So there should be:

### c) A way to handle updates for a specific path, similar to the `client.on` function.

In addition, it would be great if there was an easier way to configure a general data schema and also permission rules. Something where I can configure certain rules for certain paths. This is probably going to be a plugin, but we should make it easier to write such a plugin.

Related: #47

## Conflict resolution

The current implemented conflict handling is very primitive (if at all present): When two clients update the same path, the newest update overrides the older one. 

This has major issues right now; for example:

- Stored data: `{ posts: { 1: { title: 'What your cat does at night', likes: 15, author: 'James' } } }`
- Client A sends a delete request to `/posts/1`,
- Client B sends an update request to `/post/1/title` (values: 'Cats are weird')
  - As Client B does not know about the delete request from Client A, it will only send an update and not a complete post
- SocketDB merges the updates; it does not care about partial data
- The store does now contain: `{ posts: { 1: { title: 'Cats are weird' } } }`

This is bad, because SocketDB currently cannot decide between an update and a creation (e.g. post vs patch). SocketDB has no information about the data schema whatsoever, so there is no way to validate this.

Also, SocketDB does not currently have offline support. This results in poor user experience when the internet connection is unreliable.

To address the above mentioned issues, we need:

## To be continued..

**Comment by TimoBechtel** [comment]
This idea would solve all of the above issues: https://github.com/TimoBechtel/socketdb/issues/47#issuecomment-1633146215

**Comment by TimoBechtel** [comment]
Addressed by #79
